### PR TITLE
Fix OpenAPI Validation errors and serve ontology type meta JSON Schemas

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR makes the OpenAPI generator validate correctly, allowing us to properly generate clients.
## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202697418861671/f) _(internal)_

## 🔍 What does this change?

- Removes `Component` derive from ontology types that are too complex for compile-time reflection (all `*Type` models)
- Statically serve JSON Schemas for these ontology types
    - Added new "json_schemas` folder in the REST API with the different meta JSON Schemas
- Add mechanism for externally referring to JSON schemas from Utoipa for all models starting with `EXTERNAL_`
- Had to add anchor to PropertyType's `propertyValues` to make the resolver happy :)

## 📜 Does this require a change to the docs?

- This PR changes documentation for the API, hopefully for the better

## ⚠️ Known issues

- These added JSON Schemas are taken directly from the RFC, we have a known set of changes that need to be applied, though.
- The OpenAPI generator I've been testing against uses 

## 🐾 Next steps

- Generate Client API based on OpenAPI Spec and make use of it :tada: 

## 🛡 What tests cover this?

- None yet. I treid running [`dredd`](https://dredd.org/en/latest/) but it does not support external references.

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Start the backend and head to http://localhost:3000/api-doc/openapi.json
1.  Try to validate with `openapi-generator-cli` or other OpenAPI tool.

## 📹 Demo

image

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 1]
```diff
@@ -53,24 +63,49 @@ pub fn rest_api_router<P: GraphPool>(store: Arc<P>) -> Router {
...
+        None => Response::builder()
+            .status(StatusCode::NOT_FOUND)
+            .body(body::boxed(Empty::new()))
+            .expect("Could not create 404 response"),
+        Some(file) => Response::builder()
+            .status(StatusCode::OK)
+            .header(
+                header::CONTENT_TYPE,
+                HeaderValue::from_str("application/json")
+                    .expect("Could not create header value for found static file"),
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:92]:
```suggestion
                    .expect("could not create JSON content type header value"),
```

Considering the error is only on that

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 2]
```diff
@@ -53,24 +63,49 @@ pub fn rest_api_router<P: GraphPool>(store: Arc<P>) -> Router {
...
+            )
+            .body(body::boxed(Full::from(file.contents())))
+            .expect("Could not serve static file"),
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:95]:
```suggestion
            .expect("could not construct OK response"),
```
We're not really serving it here, right? (Although perhaps we want more than I've written, could be good to put _which_ file actually failed)

Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:95]: Do we want an `expect` here or do we actually want a `Result`? How can this fail?

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 3]
```diff
@@ -53,24 +63,49 @@ pub fn rest_api_router<P: GraphPool>(store: Arc<P>) -> Router {
...
 #[openapi(
         tags(
             (name = "Graph", description = "HASH Graph API")
         ),
-        modifiers(&MergeAddon)
+        modifiers(&MergeAddon, &ExternalRefAddon)
     )]
 struct OpenApiDocumentation;
 
+/// Addon to merge multiple API Documentations together.
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:108]: What is an "API Documentation" is there a Rust struct or something we can intra-doc link to here?

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 4]
```diff
@@ -109,3 +144,46 @@ impl Modify for MergeAddon {
         );
     }
 }
+
+/// Addon to allow external references in schemas.
+///
+/// Any component that starts with `EXTERNAL_` will transform into a relative URL in the schema and
+/// receive a `.json` ending.
+///
+/// For example the `EXTERNAL_Entity` component will be transformed into `./models/Entity.json`
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:153]: Why is it that we're mixing up casings? I understand wanting an identifiable prefix but the filenames and special-words are really ugly, especially a mix of all-caps, snake case, and pascal-case 😁 

Do we perhaps just want `external_data_type` rather than `EXTERNAL_DataType`? Or perhaps we want to go all-caps with `EXTERNAL_DATA_TYPE` or something.

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 5]
```diff
@@ -109,3 +144,46 @@ impl Modify for MergeAddon {
...
+    fn modify(&self, openapi: &mut openapi::OpenApi) {
+        for path_item in openapi.paths.paths.values_mut() {
+            for operation in path_item.operations.values_mut() {
+                if let Some(request_body) = &mut operation.request_body {
+                    modify_component_references(&mut request_body.content);
+                }
+
+                operation
+                    .responses
+                    .responses
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:166]: 🙃

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:166]: Also loving the swap to an `iter` function to avoid even more nesting 😁

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 6]
```diff
@@ -12,12 +12,22 @@ mod property_type;
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: Just shoving it here so we have a comment thread:

Is there a command-line utility we can use to validate the OpenAPI spec in automated tests and CI? It'd be good to make sure we don't accidentally invalidate it again.

I appreciate generating an OpenAPI client and using it in e2e tests might provide some confidence, so we can fall-back on that if needs be but it would be nice to validate the OpenAPI spec if we can.

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
Looking good for the most part!

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs [hunk 7]
```diff
@@ -103,7 +103,7 @@ async fn create_data_type<P: GraphPool>(
     path = "/data-types/{uri}",
     tag = "DataType",
     responses(
-        (status = 200, content_type = "application/json", description = "Data type found", body = DataType),
+        (status = 200, content_type = "application/json", description = "Data type found", body = EXTERNAL_DataType),
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:106]: Sounds good to me. Do we just update these to the singular or plural form? So simply "a data type" for this case?

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:106]: Hm, I guess if we took their one example (which they may not have intended to be thought of this deeply) as the exact way to do it then we should probably just write "A data type" yeah.

I would kind of be inclined to more go with "The requested data type" or "The schema of the requested data type" or something.

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 4 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:153]: I would have really liked if the `body` attribute supported arbitrary terms, then I'd have just put a dot instead of an underscore. The reason what just to make it clear that it's not a part of the components. I would be fine with `external_data_type`, which means having `data_type.json` in the json_schemas folder too

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:153]: I'm assuming we can't use `$data_type` or something either?

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:153]: I think the all-caps might be better to identify it as some variable/special thing. We can just `to_lowercase` in the impl and keep the file names nice

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 5 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: I rewrote these to `for in` loops but missed this one. My bad :)

Hopefully better in d19e8ae02dbb27ca4e1533c58269cdfe5466a016

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Better :) Although my 🙃 was about the `responses.responses` but we're not in control of that haha

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 6 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: The tool i used https://github.com/OpenAPITools/openapi-generator-cli seems to be one of the popular ones. It has a `validate` subcommand which I also used in the description screenshot

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: If we can throw it in the cargo make config in 10 minutes or so lets do it as part of this PR? 

Otherwise can you make an asana task for us to do so

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 4 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:153]: Addressed in ec96248ffc18bbcefff37085c9175e465b4faddb

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 6 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: Sure let's go for it. WIll do after I address other comments :)

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 3 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Addressed in 757c5e64dad3688a6fc91ef2da3fe1af7750db26

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs [hunk 8]
```diff
@@ -62,11 +62,11 @@ struct CreateDataTypeRequest {
...
     tag = "DataType",
     responses(
-      (status = 201, content_type = "application/json", description = "Data type created successfully", body = DataType),
-      (status = 422, content_type = "text/plain", description = "Provided request body is invalid"),
+        (status = 201, content_type = "application/json", description = "Data type created successfully", body = EXTERNAL_DataType),
+        (status = 422, content_type = "text/plain", description = "Provided request body is invalid"),
 
-      (status = 409, description = "Unable to create data type in the store as the base data type ID already exists"),
-      (status = 500, description = "Store error occurred"),
+        (status = 409, description = "Unable to create data type in the store as the base data type ID already exists"),
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Addressed in 87aaa98fe8031a142a5b959f821b64b64fdb2674

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Addressed in 87aaa98fe8031a142a5b959f821b64b64fdb2674

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 2 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Since this is "one level lower" than the usual routes we're creating, the `http` crate's implementation details shine through. I tried having a simpler Result return type. Let me try again to make it simpler

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: I'm asking because I didn't understand what it was 😁 (so for code comprehension rather than pub doc reasons haha)

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Thanks for clarifying

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs]: Cheers

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 2 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Fixed up in 7aebbbc10902bedb9f58551e6e2c570e5cbd37ad

Much happier with it now!

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 6 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: Addressed in cb7d5629460041d6116ec11e93952333a3dd6bb5 we might want to make it separate to `test-rest-api`, but that means spinning up the backend separately in another task

File: packages/graph/hash_graph/Makefile.toml [hunk 9]
```diff
@@ -24,11 +24,21 @@ dependencies = ["install-cargo-nextest"]
...
+trap "kill $pid" 0 ERR
+
+exit_status=0
+
 yarn httpyac send --all ${CARGO_MAKE_CURRENT_TASK_INITIAL_MAKEFILE_DIRECTORY}/tests/rest-test.http
-ret=$?
-kill "$pid"
-exit $ret
+
+# This generates a `openapitools.json` file.
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/Makefile.toml:42]: I don't understand the comment really, is it just warning there's a side-effect? Because when reading it I thought it was saying the purpose of the line

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 10]
```diff
@@ -68,31 +67,27 @@ pub fn rest_api_router<P: GraphPool>(store: Arc<P>) -> Router {
...
+            })).route("/models/*path", get(serve_static_schemas)),
         )
 }
 
 #[allow(
     clippy::unused_async,
     reason = "This route does not need async capabilities, but axum requires it in trait bounds."
 )]
-async fn static_schemas<'a>(Path(path): Path<String>) -> impl IntoResponse {
+async fn serve_static_schemas(Path(path): Path<String>) -> Result<Response, StatusCode> {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:78]: I'm not quite sure about the naming of this function, what is it actually doing? It's building a response with a given schema or something?

The actual function doesn't seem to "serve" anything, and also from how I understand it, it does not serve schema _s_, but just a specific one.

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 11]
```diff
@@ -109,3 +141,43 @@ impl Modify for MergeAddon {
...
+/// Addon to allow external references in schemas.
+///
+/// Any component that starts with `EXTERNAL_` will transform into a relative URL in the schema and
+/// receive a `.json` ending.
+///
+/// For example the `EXTERNAL_Entity` component will be transformed into `./models/Entity.json`
+struct ExternalRefAddon;
+
+impl Modify for ExternalRefAddon {
+    fn modify(&self, openapi: &mut openapi::OpenApi) {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:154]: Self requesting changes here: But we want to traverse components as well and do this substitution

Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:154]: Addressed in 1c70848b5493fd4705518f85e4ca40ac420ccdb6

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 10 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:78]: It serves the JSON Schemas we define explicitly in the `./json_schemas` folder served at `http://localhost:3000/api-doc/models/*`. 

Why doesn't it feel like it's serving anything? It serves everything in that folder for the user as a static path

File: packages/graph/hash_graph/Makefile.toml (hunk 9 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/Makefile.toml:42]: Yeah it's a side-effect. Let me rephrase it. I tried getting around the issue by adding the generated file to gitignore

Comment (thehabbos007) [reply packages/graph/hash_graph/Makefile.toml]: Addressed in b9e95694f45624b0950bdb78e11ceb1a6c6fd271

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/Makefile.toml]: We could just `rm` it but I don't feel too strongly about this

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 10 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:109]: I guess it just seems like the function is more of a builder or constructor or something. The web framework is doing the serving, this function is not. That said, I'm not really coming up with a better name 🙃 

I think we still need to figure something else out though as the name is a bit misleading, per call it is about _a_ schema, not all static schemas.

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 6 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:12]: Nice! Thanks

File: packages/graph/hash_graph/Makefile.toml (hunk 9 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/Makefile.toml]: Let me just have a look. There's a couple annoyances with this file, one being that it is created at the root and not in hash_graph.

Comment (thehabbos007) [reply packages/graph/hash_graph/Makefile.toml]: Addressed in 5fc1529c0ccfa5b5c030300147c9386ba3db1314

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs (hunk 10 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:109]: There's another project using the same setup calling this [`serve_asset` ](https://github.com/euclio/aurelius/blob/bcfafde6059efa7f55e787dd6e89759be0fb3f94/src/service.rs#L27) It seems to be of convention to call these routes something along those lines. I agree that it should be singular, though

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:109]: If we change it to singular I'm happy to approve then

Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs]: Addressed in d41bfded81b21b6d2399dc4ae3bdc9139f3b55a9

**Review by Alfred-Mountfield — APPROVED** [review]
🚀 🌠
